examples/practical-coalton/15_pathnames.lisp

```diff
@@ -0,0 +1,177 @@
+;; To load this file, you need to load Chapter 15 from the Practical
+;; Common Lisp repository. The source code can be found at the link
+;; below. Then you can copy the Chapter 15 directory into
+;; ~/quicklisp/local-projects and run (quickload "Chapter-15").
+;; https://gigamonkeys.com/book/
+
+(in-package :cl-user)
+(ql:quickload "coalton")
+(ql:quickload "Chapter-15")
+(defpackage :practical-coalton.pathnames
+  (:use
+   #:coalton
+   #:coalton-prelude)
+  (:local-nicknames
+    (#:pth #:com.gigamonkeys.pathnames))
+  (:export
+   :list-directory
+   :file-exists-p
+   :directory-pathname-p
+   :file-pathname-p
+   :pathname-as-directory
+   :pathname-as-file
+   :walk-directory
+   :directory-p
+   :file-p))
+(in-package :practical-coalton.pathnames)
+
+(named-readtables:in-readtable coalton:coalton)
+
+;; Port of Chapter 15 from Practical Common Lisp, by Peter Seibel,
+;; to Coalton.
+;;
+;; In the previous chapter, we took the approach of rewriting the
+;; Common Lisp functions in Coalton. In this chapter, we instead
+;; write a thin wrapper around the Common Lisp functions.
+;;
+;; First, we start by defining a few Coalton types that provide
+;; a type-safe interface to the Common Lisp pathname functions.
+;; For example, in several of the Common Lisp functions a wildcard
+;; pathname is an invalid argument. Instead of checking for wildcard
+;; paths and erroring if found, we can use the type system to ensure
+;; that we never pass a wildcard path to a function that doesn't
+;; accept it.
+;;
+;; Second, we define a corresponding Coalton function for each
+;; function exported by the Common Lisp pathname package.
+
+;; TODO: I'm open to any suggestions to make this code less verbose.
+;; If you count the comments at the top of the file and the package
+;; definition (which the original doesn't have), this is almost
+;; as long as the Common Lisp implementation this is wrapping!
+
+(cl:defmacro wrap-lisp-call (type function value)
+  "Wrap a call to a Lisp function that returns a value of type TYPE
+in a Coalton function that returns a value of type (Optional TYPE).
+A nil return value is converted to None."
+  `(lisp (Optional ,type) (,value)
+     (cl:let ((result (,function ,value)))
+       (cl:if result
+         (coalton (Some (lisp ,type () result)))
+         (coalton None)))))
+
+(coalton-toplevel
+
+  (define-class (Path :a))
+
+  (define-instance (Path String))
+
+  (repr :native cl:pathname)
+  (define-type Pathname)
+
+  (declare pathname->string (Pathname -> String))
+  (define (pathname->string pathname)
+    (lisp String (pathname)
+      (cl:write-to-string pathname)))
+
+  (define-instance (Path Pathname))
+
+  (repr :native cl:pathname)
+  (define-type WildPathname)
+
+  (define-instance (Path WildPathname)))
+
+(coalton-toplevel
+
+  ;; TODO: It's a little unforunate that the Either type is called
```

Comment:
It's a little unfortunate for this use-case that the Either type is called Result in Coalton, because it makes a usage to represent one of two possible values a little confusing. 

Is there a more idiomatic way to represent this?

Reply:
In this case, with a `Path` type, you get this "for free" without Either/Result/etc.

Reply:
One of the complexities in this library is there are two kinds of Lisp pathnames: wildcard pathnames and concrete pathnames. Some of the functions, like `list-directory` can only take a concrete pathname or a string. In the current implementation I'm using `(Path :a) => :a -> ...` to represent functions that can take any of the three, and `Result String Pathname` to represent functions that can only take a string or concrete pathname.

Is there a better way to encode that in the type signature?

Reply:
Ah I see what you're saying. I personally think the function ought to just error. I don't think it's worth the additional complexity just to avoid a single instance.

Another is to have more layers of types and classes to express these constraints, but I don't think that's a good idea.

Reply:
That makes sense. Thanks for the comments! I'm going to be out of town through the weekend, but I'll try to get a revision up next week.